Spin-Polarisation Selection Option

kavanase opened the pull request:
I've added a 'spin' option to the CLI interface for _sumo-dosplot_ and _sumo-bandplot_, including _sumo-bandplot_ projected BS. The syntax is explained in the docstrings and help output: "spin (:obj:`str`, optional): Plot a spin-polarised band structure, "up" for spin up only, "down" for spin down only. Defaults to ``None``."  
I've checked that it works as it should.  
Examples:  
Spin-polarised DOS (Before Changes)  
![spinpoldosorig](https://user-images.githubusercontent.com/51478689/69919592-76336e80-1476-11ea-8413-c40fea96ff09.png)  
Spin-polarised DOS, Spin Up Only (After Changes)  
![spinpoldownewspindownonly](https://user-images.githubusercontent.com/51478689/69919590-759ad800-1476-11ea-975e-a7d3f138d4b8.png)  
Spin-polarised DOS, Spin Down Only (After Changes)  
![spinpoldosnewspinuponly](https://user-images.githubusercontent.com/51478689/69919591-76336e80-1476-11ea-903b-c9763ac63fec.png)  

Spin-Polarised BS (Before Changes)  
![spinpolbandorig](https://user-images.githubusercontent.com/51478689/69919597-76cc0500-1476-11ea-968f-a16902c9fc21.png)  
Spin-Polarised BS, Spin Down Only (After Changes)  
![spinpolbandnewspindownonly](https://user-images.githubusercontent.com/51478689/69919595-76336e80-1476-11ea-993e-bb9ca956d11b.png)  
Spin-Polarised BS, Spin Up Only (After Changes)  
![spinpolbandnewspinuponly](https://user-images.githubusercontent.com/51478689/69919596-76cc0500-1476-11ea-879d-6619bdb7609c.png)  
Spin-Polarised BS with Projections (Spin Down Only)(After Changes)  
![spinpolprojbandnewspindownonly](https://user-images.githubusercontent.com/51478689/69919593-76336e80-1476-11ea-8fab-aee83de2ae2b.png)  
Spin-Polarised BS with Projections (Spin Up Only)(After Changes)  
![spinpolprojbandnewspinuponly](https://user-images.githubusercontent.com/51478689/69919594-76336e80-1476-11ea-947b-0d4ab070b23a.png)

kavanase commented:
I've also now added a flag for dosplot, bandplot and projected-bandplot, that throws up a ValueError (see below) if spin-selection called for a closed-shell calculation:
_sumo-dosplot_ for a closed-shell calculation:
![image](https://user-images.githubusercontent.com/51478689/69953889-18446c80-14f2-11ea-8fb2-27630b2ed180.png)
_sumo-bandplot_ for a closed-shell calculation:
![image](https://user-images.githubusercontent.com/51478689/69953937-2d210000-14f2-11ea-908f-823b0a4f94c0.png)
_sumo-bandplot_ (projected version) for a closed-shell calculation:
![image](https://user-images.githubusercontent.com/51478689/69953971-43c75700-14f2-11ea-8441-b84ae93929c2.png)

ajjackson commented on sumo/cli/bandplot.py:
```diff
@@ -165,6 +165,8 @@ def bandplot(filenames=None, code='vasp', prefix=None, directory=None,
 
             If ``atoms`` is not set or set to ``None`` then all atomic sites
             for all elements will be considered.
+        spin (:obj:`str`, optional): Plot a spin-polarised band structure,
```
Could this be  *"Plot only one channel of a spin-polarised ..."* to make it clearer that `None` will plot both channels?

kavanase replied:
Thanks for the helpful feedback! I've changed this for bandplot and dosplot as shown below:
![image](https://user-images.githubusercontent.com/51478689/69966003-674bcb00-150d-11ea-927f-030d391d5ac3.png)

ajjackson commented on sumo/cli/bandplot.py:
```diff
@@ -477,6 +479,8 @@ def _get_parser():
     parser.add_argument('--atoms', type=_atoms, metavar='A',
                         help=('atoms to include (e.g. "O.1.2.3,Ru.1.2.3")'))
+    parser.add_argument('--spin', type=str, default=None,
+                        help=('select spin for spin-polarised BS (options: up, down)'))
```
again, maybe "one spin channel" would be clearer?

ajjackson commented on sumo/cli/dosplot.py:
```diff
@@ -303,6 +305,8 @@ def _get_parser():
     parser.add_argument('-a', '--atoms', type=_atoms, metavar='A',
                         help=('atoms to include (e.g. "O.1.2.3,Ru.1.2.3")'))
+    parser.add_argument('--spin', type=str, default=None,
+                        help=('select spin for spin-polarised DOS (options: up, down)'))
```
as above

ajjackson commented on sumo/plotting/dos_plotter.py:
```diff
@@ -265,17 +272,30 @@ def get_plot(self, subplot=False, width=None, height=None, xmin=-6.,
             else:
                 ax = plt.gca()
 
+            ax.set_ylim(plot_data['ymin'], plot_data['ymax'])
```
Why is this happening inside a loop over lines when it only depends on the overall plot data? Shouldn't need to call this more than once.

ajjackson commented on sumo/plotting/dos_plotter.py:
```diff
@@ -265,17 +272,30 @@ def get_plot(self, subplot=False, width=None, height=None, xmin=-6.,
+                if len(spins) == 1:
+                    label = line['label']
+                    densities = line['dens'][spin][mask]
+                    ax.set_ylim(0, plot_data['ymax'])
```
I don't think it makes sense to do this repeatedly inside the loop. We could check after these loops are finished whether there were two spins and set limits accordingly?

kavanase replied:
I've moved the y-axis settings outside the loop now, and checked that it still works the same.

ajjackson commented:
This looks good to me. I'll give @utf a couple of days to chip in if he wants to and then merge.

utf commented:
Thanks @kavanase, this looks excellent.

My one suggestion would be to make the `spin` argument accept pymatgen `Spin` objects or a list of `Spin` objects rather than strings.

The command line scripts would then parse the spin option into pymatgen Spin objects (using a common function shared by both dosplot and bandplot). 

I think this route would make the implementation in the plotting functions a bit neater integrate better with the pymatgen `BandStructure` objects.

kavanase commented:
Thanks for the encouragement and suggestions @utf!
I have changed the implementation so that the parser creates a pymatgen Spin object from the cli argument (using the _string_to_spin_ function), which is then used within _bandplot_ and _dosplot_ instead of the previous string argument, as suggested. As you said, this makes the implementation a lot neater, with a couple of extraneous "if" statements and "for" loops removed as a result.

A quick summary of the changes I've made:
- Parse cli _--spin_ argument to pymatgen Spin object (also adding ValueError flag if unsuitable spin argument given)(function in _bandstructure.py_), which now allows "up", "Up", "+1" or "1" for spin up, "down", "Down" or "-1" for spin down.
- Moved "if" statement outside "for" loop in _bs_plotter_ (neater and more efficient).
- Removed several "if" statements via the implementation of Spin object instead of string (_bs_plotter_ & _dos_plotter_)
- Removed an extra 'for' loop via implementation of Spin rather than string (_bs_plotter_) 
- One or two bits of reformatting to make implementation slightly neater.
- Removed an "elif" statement via the implementation of Spin object instead of string (_dos_plotter_) 
- Updated docstrings to reflect slightly different implementation, and the fact that spin can also now be specified with "1" or "-1".

Again, I've quickly checked that _sumo-dosplot_, _sumo-bandplot_ and _sumo-bandplot_ _project_ all still work as intended; some examples:  
Error flag for incorrect spin argument:  
![image](https://user-images.githubusercontent.com/51478689/70184336-8d28c980-16df-11ea-9e75-21ecab1be123.png)

_sumo-bandplot_
![1909bandpoltest](https://user-images.githubusercontent.com/51478689/70184384-a2055d00-16df-11ea-9e52-b5d2662b9a92.png)
_sumo-bandplot --spin -1_
![1908banddowntest](https://user-images.githubusercontent.com/51478689/70184385-a2055d00-16df-11ea-806b-eb43667551ba.png)
_sumo-bandplot --project Br.p, Ag.d --spin up_
![1948bandprojuptest](https://user-images.githubusercontent.com/51478689/70184383-a16cc680-16df-11ea-8396-7240b71c3524.png)
_sumo-dosplot_
![1959dospoltest](https://user-images.githubusercontent.com/51478689/70184388-a2055d00-16df-11ea-835a-d747ca571a63.png)
_sumo-dosplot --spin down_
![2000dosdowntest](https://user-images.githubusercontent.com/51478689/70184386-a2055d00-16df-11ea-933a-b37e1d4fbf55.png)

_sumo-dosplot --spin 1_
![1959dosuptest](https://user-images.githubusercontent.com/51478689/70184390-a29df380-16df-11ea-9277-7d2d8a731290.png)

utf commented:
Fantastic! This looks great.

Thanks for making those changes and being so thorough with the testing. This will be a very useful feature.
